platform/engineering/design-docs/micro-frontend-architecture.md

```diff
@@ -0,0 +1,255 @@
+# Micro Front end Architecture
+
+**Author(s):** Chris Valarida (chris.valarida@adhocteam.us)  
+**Last Updated:** February XX, 2020  
+**Status:** **Draft** | In Review | Approved  
+**Approvers:** Andrew Gunsch, Rian Fowler, Wyatt Walter
+
+<!-- Should the approvers include anybody from DEPO? -->
+
+
+## Overview
+
+### Objective
+To establish a micro front end architecture which enables application teams to
+operate independently.
+
+This document is intended for front end and DevOps engineers on the
+Veteran-facing Services Platform (VSP) and Veteran-Facing Services (VSA) teams.
+
+
+### Background
+
+<!-- Could put a definitions section here if it becomes useful -->
+
+#### The State of `vets-website`
+
+##### Build
+The `vets-website` build is responsible for a lot of things. The most important
+pieces include:
+- Pulling content from the [CMS](#CMS)
+- Building that content into HTML pages
+- Validating the content for broken links, accessibility violations, etc.
+- Building all JavaScript bundles
+  - Both the common bundle and individual application bundles
+- Building all CSS bundles
+  - Both common and application-specific
+- Running all >5k unit tests for all applications
+- Running all end-to-end tests for all applications
+
+
+##### Deploy
+When `vets-website` is deployed, all content and applications are deployed
+together. **There is no way to deploy one application without deploying
+everything.**
+
+There is, however, a way to deploy the latest content from the CMS without
+deploying the latest application code. This is called a **content-only deploy**.
+The latest `vets-website` release is used instead of building code in the latest
+commit of the `vets-website` `master` branch.
+
+
+#### The Problems
+##### Failures in one Service affect all other Services
+When one application team introduces test failures into `master`, all PRs from
+all teams that include the offending commit in `master` will fail in Continuous
+Integration (CI) because all tests are run, not just the tests for a single
+application.
+
+**Example:** Team A introduces a test failure in `master`. Team B has an urgent
+bug fix that needs deployed as soon as possible, so they make a branch of
+`master` and quickly fix it. Team B's branch can't be merged until Team A's
+unrelated test failure is fixed.
+
+Likewise, when there's a content issue, such as an accessibility violation, all
```

Comment:
same question, how often does this happen? what leads to it occurring?

Reply:
Back before broken links in a build triggered a fallback to the cached content, it was as much as three or four times times a week. After we made that change, it's been pretty rare. I'd guess maybe three or four times in the past four months or so?

I don't know what all the avenues are that lead to this happening, but last time, it was because a content editor entered invalid HTML into a WYSIWYG component. I think it was forgetting to close a `<ul>`.

As we get more and more content editors on board, the chances of this happening only increase. I think the CMS team mentioned a safety measure they're going to put in place, for what it's worth.

Reply:
if it's from bad CMS content, this seems like the kind of thing that should be improved by CMS fixes to their own CI (or FE build), not rearchitecting our entire repo + deployment process :)

if the other things are "once a month", same comment. they're nice things to fix, but they don't justify this level of overhaul.